File: frameworks/python/fastapi/Implementation/routes/authorized.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException
from fastapi.security import HTTPAuthorizationCredentials, HTTPBearer

from payloads import Payload, Payloads


def router(p: Payloads) -> APIRouter:
    """authorized: FastAPI's bearer scheme reads the token, and a dependency on the route compares it
    before the handler runs."""
    token = p.settings.token

    # rb:wiring authorized.*
    bearer = HTTPBearer()

    async def require_token(credentials: Annotated[HTTPAuthorizationCredentials, Depends(bearer)]) -> None:
        """A request with no bearer token is HTTPBearer's own 401. Any token but settings.json's is 403."""
        if credentials.credentials != token:
            raise HTTPException(status_code=403)
    # rb:end

    routes = APIRouter()

    @routes.get("/authorized/small", dependencies=[Depends(require_token)])
    async def small() -> Payload:
        return p.small

    return routes

File: frameworks/rust/actix-web/Implementation/routes/authorized.rs
use actix_web::dev::ServiceRequest;
use actix_web::error::ErrorForbidden;
use actix_web::web::{self, ServiceConfig};
use actix_web_httpauth::extractors::bearer::BearerAuth;
use actix_web_httpauth::middleware::HttpAuthentication;

use crate::Payloads;

/// authorized: actix-web-httpauth's bearer middleware on this one resource. It reads the bearer
/// token from the Authorization header before the handler runs and hands it to the validator,
/// which refuses any token but settings.json's with 403.
pub fn configure(cfg: &mut ServiceConfig, p: &'static Payloads) {
    // rb:wiring authorized.*
    let bearer = HttpAuthentication::bearer(move |request: ServiceRequest, credentials: BearerAuth| async move {
        if credentials.token() == p.settings.token { Ok(request) } else { Err((ErrorForbidden("the bearer token is not the one this route accepts"), request)) }
    });

    cfg.service(web::resource("/authorized/small").wrap(bearer).route(web::get().to(move || async move { web::Json(&p.small) })));
}
